pkg/clustermanager/kubetest2/gke.go

```diff
@@ -0,0 +1,173 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package kubetest2
+
+import (
+	"fmt"
+	"log"
+	"os/exec"
+	"regexp"
+	"strconv"
+	"strings"
+
+	"knative.dev/test-infra/pkg/cmd"
+	"knative.dev/test-infra/pkg/metautil"
+	"knative.dev/test-infra/pkg/prow"
+)
+
+const (
+	createCommandTmpl = "%s container clusters create --quiet --enable-autoscaling --min-nodes=%d --max-nodes=%d " +
+		"--cluster-version=%s --scopes=%s --no-enable-autoupgrade"
+	boskosAcquireDefaultTimeoutSeconds = 1200
+)
+
+var (
+	baseKubetest2Flags = []string{"gke", "--ignore-gcp-ssh-key=true", "--up"}
+
+	// If one of the error patterns below is matched, it would be recommended to
+	// retry creating the cluster in a different region.
+	// - stockout (https://github.com/knative/test-infra/issues/592)
+	retryableCreationErrors = []*regexp.Regexp{
+		regexp.MustCompile(`.*does not have enough resources available to fulfill.*`),
+		regexp.MustCompile(`.*only \d+ nodes out of \d+ have registered; this is likely due to Nodes failing to start correctly.*`),
+	}
+)
+
+// GKEClusterConfig are the supported configurations for creating a GKE cluster.
+type GKEClusterConfig struct {
+	GCPProjectID                string
+	BoskosAcquireTimeoutSeconds int
+	Name                        string
+	Region                      string
+	BackupRegions               []string
+	Machine                     string
+	MinNodes                    int
+	MaxNodes                    int
+	Network                     string
+	Version                     string
+	Scopes                      string
+	Addons                      string
+	Environment                 string
+	CommandGroup                string
+	PrivateClusterAccessLevel   string
+	PrivateClusterMasterIPRange string
+}
+
+// Run will run the `kubetest2 gke` command with the provided parameters,
+// it will also handle the logic that is only used for Knative integration testing, like retrying cluster creation.
+func Run(opts *Options, cc *GKEClusterConfig) error {
+	createCommand := fmt.Sprintf(createCommandTmpl, cc.CommandGroup, cc.MinNodes, cc.MaxNodes, cc.Version, cc.Scopes)
+	if cc.Addons != "" {
+		createCommand += " --addons=" + cc.Addons
+	}
+	kubetest2Flags := append(baseKubetest2Flags, "--create-command="+createCommand)
+
+	kubetest2Flags = append(kubetest2Flags, "--cluster-name="+cc.Name, "--environment="+cc.Environment,
+		"--num-nodes="+strconv.Itoa(cc.MinNodes), "--machine-type="+cc.Machine, "--network="+cc.Network)
+
+	if prow.IsCI() && cc.GCPProjectID == "" {
+		log.Println("Will use boskos to provision the GCP project")
+		timeout := cc.BoskosAcquireTimeoutSeconds
+		if timeout == 0 {
+			timeout = boskosAcquireDefaultTimeoutSeconds
+		}
+		kubetest2Flags = append(kubetest2Flags, "--boskos-acquire-timeout-seconds="+strconv.Itoa(timeout))
+	} else {
+		if cc.GCPProjectID == "" {
+			var err error
+			cc.GCPProjectID, err = cmd.RunCommand("gcloud config get-value project")
```

Comment:
This should be a documented behavior. Also, do we want to have it?

Reply:
I've seen some brief explanation below, but this can have undesired consequences if not known. I would remove this backup case unless there is a strong requirement including it.

Reply:
This is the default behavior for `gcloud` command - if the command does not provide a project, e.g. `gcloud container clusters create test`, it will use the current GCP project, I think it's better to follow the same pattern? (This is mostly for local usage, so there is not a strong requirement to include it but I feel it does not hurt much)

Reply:
This follows the same guidance as the comment above. We're using gcloud, but ideally our customers should not care or know how we're fulfilling their requests. I'd rather implement a behavior that is holistically consistent and independent from gcloud in its definition.
What do you think?

Reply:
Makes perfect sense. I have updated the logic to throw out an error if users do not provide the GCP project ID in the non-CI environment.